fizz/crypto/experimental/hacl/Hacl.cpp

```diff
@@ -0,0 +1,105 @@
+#include "Hacl.h"
+
+#include "Hacl_AesGCM_NI.h"
+#include <fizz/crypto/aead/IOBufUtil.h>
+#include <folly/Range.h>
+#include <folly/lang/Bits.h>
+
+namespace fizz {
+namespace hacl {
+
+std::array<uint8_t, 12> Hacl::createIV(uint64_t seqNum) const {
+  std::array<uint8_t, 12> iv;
+  uint64_t bigEndianSeqNum = folly::Endian::big(seqNum);
+  const size_t prefixLength = 12 - sizeof(uint64_t);
+  memset(iv.data(), 0, prefixLength);
+  memcpy(iv.data() + prefixLength, &bigEndianSeqNum, 8);
+  XOR(key_.iv->coalesce(), folly::range(iv));
+  return iv;
+}
+
+void Hacl::setKey(TrafficKey tk) {
+  tk.key->coalesce();
+  tk.iv->coalesce();
+  key_ = std::move(tk);
+}
+
+std::unique_ptr<folly::IOBuf> Hacl::encrypt(
+    std::unique_ptr<folly::IOBuf>&& plaintext,
+    const folly::IOBuf* associatedData,
+    uint64_t seqNum) const {
+  Lib_Vec128_vec128 ctx[22] = {0};
+  // get iv and init hacl
+  auto iv = createIV(seqNum);
+  uint8_t* keyData = const_cast<uint8_t*>(key_.key->data());
+  Hacl_AesGCM_NI_aes128_gcm_init(ctx, keyData, iv.data());
```

Comment:
do you need to modify the key here? or is it just a limitation of the hacl interface?

iobuf has a writableData() as well which will give you a non const ptr to the data

Reply:
const cast is fine if you dont have to change the key but its just takes a non const ptr